File: dotnet/src/DuckDuckGo/Infrastructure/WebDriverFactory.cs
using System;
using OpenQA.Selenium;
using OpenQA.Selenium.Chrome;
using OpenQA.Selenium.Edge;
using OpenQA.Selenium.Firefox;
using OpenQA.Selenium.IE;
using OpenQA.Selenium.Remote;

namespace DuckDuckGo.Infrastructure
{
    internal static class WebDriverFactory
    {
        private static readonly NLog.Logger Logger = NLog.LogManager.GetCurrentClassLogger();

        private static IWebDriver _driver;
        private static readonly Uri RemoteUrl = new Uri(@"https://localhost:4444/hub");

        internal static IWebDriver CreateWebDriver(Browser browser, bool headless = false)
        {
            switch (browser)
            {
                case Browser.Chrome:
                    _driver = CreateDefaultChromeDriver(headless);
                    break;

                case Browser.Edge:
                    _driver = new EdgeDriver();
                    break;

                case Browser.InternetExplorer:
                    _driver = new InternetExplorerDriver();
                    break;

                case Browser.Remote:
                    var options = new FirefoxOptions();
                    options.PlatformName = "Linux";
                    _driver = new RemoteWebDriver(RemoteUrl, options.ToCapabilities());
                    break;

                default:
                    throw new ArgumentException("Browser is not supported or valid!");
            }

            LogBrowserVersion(_driver);
            _driver.Manage().Cookies.DeleteAllCookies();
            return _driver;
        }

        private static IWebDriver CreateDefaultChromeDriver(bool headless)
        {
            var service = ChromeDriverService.CreateDefaultService();
            service.HideCommandPromptWindow = true;
            service.EnableVerboseLogging = false;
            service.SuppressInitialDiagnosticInformation = true;
            service.Start();

            var options = new ChromeOptions();
            options.AddArgument("start-maximized");

            if (headless)
            {
                options.AddArgument("headless");
                options.AddArgument("disable-gpu");
            }

            _driver = new ChromeDriver(service, options);
            return _driver;
        }

        internal static void DestroyWebDriver()
        {
            if (_driver != null)
            {
                _driver.Quit();
                _driver.Dispose();
            }

            _driver = null;
        }

        private static void LogBrowserVersion(IWebDriver driver)
        {
            if (driver is IHasCapabilities hasCapabilities)
            {
                var capabilities = hasCapabilities.Capabilities;
                Logger.Debug("Browser=" + capabilities.GetCapability("browserName") +
                             "Version=" + capabilities.GetCapability("browserVersion") +
                             "Platform=" + capabilities.GetCapability("platformName"));
            }
        }
    }
}

File: java/src/test/java/com/sqs/selenium/playground/infrastructure/WebDriverFactory.java
package com.sqs.selenium.playground.infrastructure;

import com.sqs.selenium.playground.infrastructure.exceptions.UnsupportedBrowserException;
import org.openqa.selenium.Platform;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.chrome.ChromeDriver;
import org.openqa.selenium.chrome.ChromeOptions;
import org.openqa.selenium.remote.DesiredCapabilities;
import org.openqa.selenium.remote.RemoteWebDriver;

import java.net.MalformedURLException;
import java.net.URL;

public class WebDriverFactory {

    private static WebDriver driver;
    private static URL remote;

    public enum Browser {
        HEADLESS,
        CHROME,
        FIREFOX
    }

    /**
     * Factory method that returns a WebDriver instance
     * based on the given browser
     *
     * @param browser enum representing the local browser
     * @return WebDriver instance
     */

    public static WebDriver getInstance(Browser browser) {

        if (browser == null) {
            browser = Browser.CHROME;
        }

        try {
            remote = UrlBuilder.empty()
                    .withProtocol("http")
                    .withHost("localhost:4444")
                    .withPath("/hub")
                    .build();
        } catch (MalformedURLException e) {
            throw new IllegalArgumentException("Obviosuly not a valid Grid URL!");
        }

        switch (browser) {
            case CHROME:
                driver = new ChromeDriver();
                break;
            case HEADLESS:
                final ChromeOptions options = new ChromeOptions();
                options.addArguments("--headless");
                options.addArguments("--disable-gpu");
                driver = new ChromeDriver(options);
                break;
            case FIREFOX:
                DesiredCapabilities capabilities = DesiredCapabilities.firefox();
                capabilities.setPlatform(Platform.LINUX);
                capabilities.setBrowserName("Firefox");
                driver = new RemoteWebDriver(remote, capabilities);
            default:
                throw new UnsupportedBrowserException("Unsupported browser," + browser);
        }
        driver.manage().deleteAllCookies();
        return driver;
    }
}